perllib/FixMyStreet/Roles/DDProcessor.pm

```diff
@@ -0,0 +1,367 @@
+                $r->state('hidden');
+                $r->update;
+            }
+        } else {
+            # We don't do anything with DD cancellations that don't have
+            # associated Cancel reports, so no need to warn on them
+            # warn "no matching record found for Cancel payment with id $payer\n";
+        }
+        $self->log("finished looking at payment " . $payment->{$self->cancelReferenceField});
+    }
+}
+
+sub add_new_sub_metadata { return; }
+
+sub _report_matches_payment {
+    my ($self, $r, $p) = @_;
+
+    my $match = 0;
+    if ( $p->{$self->oneOffReferenceField} && $r->id eq $p->{$self->oneOffReferenceField} ) {
+        $match = 'Ad-Hoc';
+    } elsif ( !$p->{$self->oneOffReferenceField}
+            && ( $r->get_extra_field_value('Subscription_Type') eq $self->waste_subscription_types->{New} ||
+                 # if we're renewing a previously non DD sub
+                 $r->get_extra_field_value('Subscription_Type') eq $self->waste_subscription_types->{Renew} )
+    ) {
+        $match = 'New';
+    }
+
+    return $match;
+}
+
+sub _duplicate_waste_report {
+    my ( $report, $category, $extra ) = @_;
+    my $new = FixMyStreet::DB->resultset('Problem')->new({
+        category => $category,
+        user => $report->user,
+        latitude => $report->latitude,
+        longitude => $report->longitude,
+        cobrand => $report->cobrand,
+        bodies_str => $report->bodies_str,
+        title => $report->title,
+        detail => $report->detail,
+        postcode => $report->postcode,
+        used_map => $report->used_map,
+        name => $report->user->name || $report->name,
+        areas => $report->areas,
+        anonymous => $report->anonymous,
+        state => 'unconfirmed',
+        non_public => 1,
+    });
+
+    $extra->{Subscription_Details_Container_Type} ||= $report->get_extra_field_value('Subscription_Details_Container_Type');
+    $extra->{service_id} ||= $report->get_extra_field_value('service_id');
+
+    my @extra = map { { name => $_, value => $extra->{$_} } } keys %$extra;
+    $new->set_extra_fields(@extra);
+
+    return $new;
+}
+
+sub waste_get_current_garden_sub {
+    my ( $self, $id ) = @_;
+
+    my $echo = $self->feature('echo');
+    $echo = Integrations::Echo->new(%$echo);
+    my $services = $echo->GetServiceUnitsForObject( $id );
+    return undef unless $services;
+
+    my $garden;
+    for my $service ( @$services ) {
+        if ( $service->{ServiceId} == $self->garden_waste_service_id ) {
+            $garden = $self->_get_current_service_task($service);
+            last;
+        }
+    }
+
+    return $garden;
+}
+
+sub waste_get_sub_quantity {
+    my ($self, $service) = @_;
+
+    my $quantity = 0;
+    my $tasks = Integrations::Echo::force_arrayref($service->{Data}, 'ExtensibleDatum');
+    return 0 unless scalar @$tasks;
+    for my $data ( @$tasks ) {
+        next unless $data->{DatatypeName} eq 'LBB - GW Container';
```

Comment:
This is Bromley specific text here. (SLWP - Containers for Kingston, used in cobrand elsewhere, so we should give it its own cobrand lookup I assume).